### ISSUE TITLE
removing nfs volume hangs

### ISSUE THREAD

**Issue by vikstrous** [moby/moby#31390]
**Description**

We've been seeing an intermittent race when removing NFS volumes that causes the engine to hang for half an hour and not be able to run volume remove or volume list operations (and possibly others I haven't tested). The issue happens when the NFS server is down during an NFS volume remove. I suspect that the issue is triggered when the volume remove is done immediately after a container remove. In my repro I set up two volumes with two containers to increase the chance of triggering the issue, but I'm not completely sure what the bare minimum is for this repro.

**Steps to reproduce the issue:**

(Start NFS server)

```
docker volume create -o type=nfs -o o=addr=127.0.0.1,rw,sync,actimeo=0 -o device=:/home/v/nfsshare nfsvol
docker volume create -o type=nfs -o o=addr=127.0.0.1,rw,sync,actimeo=0 -o device=:/home/v/nfsshare nfsvol2
```

```
docker run --rm -it --name contianer1 -v nfsvol:/nfs alpine sh
docker run --rm -it --name container2 -v nfsvol2:/nfs alpine sh
```

(Stop NFS server)

```
docker rm -f container1 container2; docker volume rm nfsvol nfsvol2
container2
Error response from daemon: No such container: container1
Error response from daemon: Unable to remove volume, volume still in use: remove nfsvol: volume is in use - [8614e37d79038d20f4ecf11f2974ca5ffa55631935478d7f5eb76fa6a9ae5e58]
^[[A
```

The volume rm command hangs.

Sorry, I made a typo in this repro so it's not actually deleting both containers, but I'm not sure which parts are significant so I'll leave it as it is for now. The issue is not 100% reproducible so it's not easy to keep re-testing it.

**Describe the results you received:**
The volume rm command hangs.

**Describe the results you expected:**
The volume rm command succeeds.

**Additional information you deem important (e.g. issue happens only occasionally):**
Intermittent issue

**Output of `docker version`:**

```
Client:
 Version:      1.14.0-dev
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   422118c-unsupported
 Built:        Tue Feb  7 06:17:58 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.14.0-dev
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   422118c-unsupported
 Built:        Tue Feb  7 06:17:58 2017
 OS/Arch:      linux/amd64
 Experimental: false
```
Sorry, I'm running my own fork, but this has been happening on docker 1.13.1 as well.

**Output of `docker info`:**

```
Containers: 13
 Running: 11
 Paused: 0
 Stopped: 2
Images: 2024
Server Version: 1.14.0-dev
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: d39luhauv7cd9n1xm8dv2vjvc
 Is Manager: true
 ClusterID: rjhxneh865hstihwuhecg5krp
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  External CAs:
    cfssl: https://172.17.0.1:12381/api/v1/cfssl/sign
 Node Address: 172.17.0.1
 Manager Addresses:
  172.17.0.1:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 78fb8f45890a601e0fd9051cf9f9f74923e950fd
runc version: 51371867a01c467f08af739783b8beafc154c4d7
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.8.17
Operating System: NixOS 17.03.git.6378f62 (Gorilla)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.53 GiB
Name: thisisfine
ID: 4XB6:RAOL:2UT7:W3TT:JBSR:ILHC:LREA:HUVU:5U7V:TI7E:WUT2:WQTS
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 108
 Goroutines: 192
 System Time: 2017-02-27T09:40:47.981254363-08:00
 EventsListeners: 2
Username: viktorstanchev
Registry: https://index.docker.io/v1/
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Experimental: false
Insecure Registries:
 172.17.0.1:3001
 172.17.0.1:443
 172.17.0.1:444
 172.17.0.1:445
 172.17.0.1
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical

**Comment by cpuguy83** [comment]
This seems to be caused primarily by the call to `unmount` (which is called when the container exits) hanging when the nfs server is down.

Then when `Remove` is called it's trying to stat/eval symlinks on the volume path, which is still mounted. This also hangs since it's trying to stat against the nfs endpoint.

The issue with `unmount` hanging seems like it can be fixed by unmounting with `MNT_DETACH`

However the `Remove` call on the local driver could likely do more to ensure this doesn't happen as well... like either return an error or attempt a force unmount if mounted.

**Comment by cpuguy83** [comment]
Fix for the the `unmount` hang is in #31391

**Comment by cpuguy83** [comment]
Fixed by #31391 and #32327
